### DISCUSSION TITLE
Need implicit type convert in "is" constant pattern matching

### DISCUSSION THREAD

**Discussion by ahdung** [dotnet/csharplang#4401] [Language Ideas]
this code `obj is 0`, if i'm not sure the `obj`'s type is `int`, then i dare not write like this, because the result is probably not as expected.
```c#
byte a = 0;
object obj = a;
obj is 0 // false
```
This limits the use of `is`, so i hope it do type convert.

**Comment by alrz** [comment]
Try `obj is (byte)0`. 

You need the same type when you are unboxing. `0` is `int`..

**Reply by ahdung** [reply]
I understand the reason is not do type convert when using `is`, but/so I hope it do. Then, when I only care about whether `obj` is greater than 0, I can just write this directly:
```c#
if (obj is >0) {
    DoSomething();
}
```
instead of:
```c#
if    (obj is int a && a > 0
    || obj is byte b && b > 0
    || obj is short c && c > 0
    || obj is long d && d > 0
    || obj is sbyte e && e > 0
    || obj is ushort f && f > 0
    || obj is uint g && g > 0
    || obj is ulong h && h > 0
    || obj is float i && i > 0
    || obj is double j && j > 0
    || obj is decimal k && k > 0
    )
{
    DoSomething();
}
```

**Reply by alrz** [reply]
That's what the compiler would have to do anyways if it was going to work, which is just wasteful if you know what type you expect. You could simplify that to `obj is int and > 0 or byte and > 0` etc but you have to mention every type.

At the moment C# lacks generic operations on numerics, see https://github.com/dotnet/csharplang/discussions/164.

**Reply by ahdung** [reply]
When write something about UI with formatting, this situation is quite common, I don't know the data source give me what, I just care whether the value meets certain conditions.

**Reply by CyrusNajmabadi** [reply]
you could use Convert or IConvertible to attempt to convert it to an int first, then check the value after that.

**Reply by ahdung** [reply]
explicit Convert might cause exception, like `DBNull`, it's not a good option.

**Reply by CyrusNajmabadi** [reply]
Then you'll have to write an extension that tests out all the cases.

**Reply by 333fred** [reply]
> explicit Convert might cause exception, like `DBNull`, it's not a good option.

We couldn't automatically generate anything that worked better. You're just including the default language built-in numeric types here. You could have a `Rational` number type, or an `Imaginary` number type, and checking them for being 0 would also be valid. But we have no way of generating that code.

**Reply by ahdung** [reply]
@333fred In my opinion, just implicit convert `obj` to the type of this constant(this case is `int`) is enough. If the conversion fails, just return `false`.

**Reply by svick** [reply]
@ahdung That's still not possible. To perform an implicit cast, the compiler needs to know the the type to cast from. Or it would have to do something similar to reflection or `dynamic`, but those are generally not acceptable without explicit opt-in.

**Reply by ahdung** [reply]
Anyway, `obj is 0` return `false` just because `obj` is not an `int`, this is not graceful.

**Reply by huoyaoyuan** [reply]
The limitation comes from type system, not pattern matching.

**Comment by HaloFour** [comment]
This feels like perhaps a use case for active patterns?

**Comment by theunrepentantgeek** [comment]
Rereading this thread, I was struct by something in this block:

``` csharp
if    (obj is int a && a > 0
    || obj is byte b && b > 0
    || obj is short c && c > 0
    || obj is long d && d > 0
    || obj is sbyte e && e > 0
    || obj is ushort f && f > 0
    || obj is uint g && g > 0
    || obj is ulong h && h > 0
    || obj is float i && i > 0
    || obj is double j && j > 0
    || obj is decimal k && k > 0
    )
{
    DoSomething();
}
```

Across all these _different_ types of numbers (different ranges, precisions, etc), comparisons with zero (0) are about the only meaningful common operation. 

Addition and subtraction can't be generalized because of the differences in range (code that could never overflow with **int** is extremely likely to with **byte**, for example). Same for multiplication.

Division is a whole 'nother can of worms - some of the types support floating point division, others only integer based.

Even if we restrict ourselves to just comparison operators, things don't work.

Let's imagine for a moment that this desired code worked:

``` csharp
if (obj is < 6) {
    DoSomething();
}
```
(The only differences are using `<` and changing the boundary to `6`)

If `obj` contained `5.7` (a **double** value) then implicit conversion can't work because there's no implicit conversion from double to int. 

If it was forced (say, using `Convert.ToInt32(5.7)`) then the answer would be wrong, with the condition returning **false** (because `Convert.ToInt32(5.7) == 6`).

I don't think relying on implicit conversions is a good idea.

Instead, within the current bounds of C#, and assuming that `obj` does contain some kind of numeric type, you can convert it pretty easily:

``` csharp
var value = (double)Convert.ToDouble((dynamic)obj);
```

Using **dynamic** forces overload resolution to happen at runtime (instead of compile time), so any of the numeric types supported by `Convert.ToDouble()` will be handled cleanly. 

Issues to be aware of include performance (**dynamic** is slow, but not that slow, especially if you're working with human perceptible speeds), and exceptions (if `obj` is not a numeric type, thing will blow up).
